File: petclinic-java-gradle-react-tailwind-ts/src/main/java/dev/pollito/petclinic_java_gradle_react_tailwind_ts/pet/PetServiceImpl.java
package dev.pollito.petclinic_java_gradle_react_tailwind_ts.pet;

import dev.pollito.petclinic_java_gradle_react_tailwind_ts.events.BeforeDeleteOwner;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.events.BeforeDeletePet;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.events.BeforeDeleteType;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.owner.OwnerRepository;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.type.TypeRepository;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.util.CustomCollectors;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.util.NotFoundException;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.util.ReferencedException;
import java.util.List;
import java.util.Map;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.context.event.EventListener;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

@Service
public class PetServiceImpl implements PetService {

    private final PetRepository petRepository;
    private final TypeRepository typeRepository;
    private final OwnerRepository ownerRepository;
    private final ApplicationEventPublisher publisher;
    private final PetMapper petMapper;

    public PetServiceImpl(final PetRepository petRepository, final TypeRepository typeRepository,
            final OwnerRepository ownerRepository, final ApplicationEventPublisher publisher,
            final PetMapper petMapper) {
        this.petRepository = petRepository;
        this.typeRepository = typeRepository;
        this.ownerRepository = ownerRepository;
        this.publisher = publisher;
        this.petMapper = petMapper;
    }

    @Override
    public List<PetDTO> findAll() {
        final List<Pet> pets = petRepository.findAll(Sort.by("id"));
        return pets.stream()
                .map(pet -> petMapper.updatePetDTO(pet, new PetDTO()))
                .toList();
    }

    @Override
    public PetDTO get(final Integer id) {
        return petRepository.findById(id)
                .map(pet -> petMapper.updatePetDTO(pet, new PetDTO()))
                .orElseThrow(NotFoundException::new);
    }

    @Override
    public Integer create(final PetDTO petDTO) {
        final Pet pet = new Pet();
        petMapper.updatePet(petDTO, pet, typeRepository, ownerRepository);
        return petRepository.save(pet).getId();
    }

    @Override
    public void update(final Integer id, final PetDTO petDTO) {
        final Pet pet = petRepository.findById(id)
                .orElseThrow(NotFoundException::new);
        petMapper.updatePet(petDTO, pet, typeRepository, ownerRepository);
        petRepository.save(pet);
    }

    @Override
    public void delete(final Integer id) {
        final Pet pet = petRepository.findById(id)
                .orElseThrow(NotFoundException::new);
        publisher.publishEvent(new BeforeDeletePet(id));
        petRepository.delete(pet);
    }

    @Override
    public Map<Integer, Integer> getPetValues() {
        return petRepository.findAll(Sort.by("id"))
                .stream()
                .collect(CustomCollectors.toSortedMap(Pet::getId, Pet::getId));
    }

    @EventListener(BeforeDeleteType.class)
    public void on(final BeforeDeleteType event) {
        final ReferencedException referencedException = new ReferencedException();
        final Pet typePet = petRepository.findFirstByTypeId(event.getId());
        if (typePet != null) {
            referencedException.setKey("type.pet.type.referenced");
            referencedException.addParam(typePet.getId());
            throw referencedException;
        }
    }

    @EventListener(BeforeDeleteOwner.class)
    public void on(final BeforeDeleteOwner event) {
        final ReferencedException referencedException = new ReferencedException();
        final Pet ownerPet = petRepository.findFirstByOwnerId(event.getId());
        if (ownerPet != null) {
            referencedException.setKey("owner.pet.owner.referenced");
            referencedException.addParam(ownerPet.getId());
            throw referencedException;
        }
    }

}

File: petclinic-kt-gradle-thymeleaf-tailwind-webpack-htmx-ts/src/main/kotlin/dev/pollito/petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts/pet/PetServiceImpl.kt
package dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.pet

import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.events.BeforeDeleteOwner
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.events.BeforeDeletePet
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.events.BeforeDeleteType
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.owner.OwnerRepository
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.type.TypeRepository
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.util.CustomCollectors
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.util.NotFoundException
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.util.ReferencedException
import org.springframework.context.ApplicationEventPublisher
import org.springframework.context.event.EventListener
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service


@Service
class PetServiceImpl(
    private val petRepository: PetRepository,
    private val typeRepository: TypeRepository,
    private val ownerRepository: OwnerRepository,
    private val publisher: ApplicationEventPublisher,
    private val petMapper: PetMapper
) : PetService {

    override fun findAll(): List<PetDTO> {
        val pets = petRepository.findAll(Sort.by("id"))
        return pets.map { pet -> petMapper.updatePetDTO(pet, PetDTO()) }
    }

    override fun `get`(id: Int): PetDTO = petRepository.findById(id)
            .map { pet -> petMapper.updatePetDTO(pet, PetDTO()) }
            .orElseThrow { NotFoundException() }

    override fun create(petDTO: PetDTO): Int {
        val pet = Pet()
        petMapper.updatePet(petDTO, pet, typeRepository, ownerRepository)
        return petRepository.save(pet).id!!
    }

    override fun update(id: Int, petDTO: PetDTO) {
        val pet = petRepository.findById(id)
                .orElseThrow { NotFoundException() }
        petMapper.updatePet(petDTO, pet, typeRepository, ownerRepository)
        petRepository.save(pet)
    }

    override fun delete(id: Int) {
        val pet = petRepository.findById(id)
                .orElseThrow { NotFoundException() }
        publisher.publishEvent(BeforeDeletePet(id))
        petRepository.delete(pet)
    }

    override fun getPetValues(): Map<Int, Int> = petRepository.findAll(Sort.by("id"))
            .stream()
            .collect(CustomCollectors.toSortedMap(Pet::id, Pet::id))

    @EventListener(BeforeDeleteType::class)
    fun on(event: BeforeDeleteType) {
        val referencedException = ReferencedException()
        val typePet = petRepository.findFirstByTypeId(event.id)
        if (typePet != null) {
            referencedException.key = "type.pet.type.referenced"
            referencedException.addParam(typePet.id)
            throw referencedException
        }
    }

    @EventListener(BeforeDeleteOwner::class)
    fun on(event: BeforeDeleteOwner) {
        val referencedException = ReferencedException()
        val ownerPet = petRepository.findFirstByOwnerId(event.id)
        if (ownerPet != null) {
            referencedException.key = "owner.pet.owner.referenced"
            referencedException.addParam(ownerPet.id)
            throw referencedException
        }
    }

}
